### ISSUE TITLE
Each segment being parsed twice

### ISSUE THREAD

**Issue by AFAde** [duaneedwards/nHapi#3]
Hi all,
I have noticed that each segment is being parsed twice. Normally this is not a big issue but if the hl7 file contains a big embedded PDF as base64 string, it actually causes serious performance issues.

Are you aware of this problem? If needed I can provide more info about that.

**Comment by duaneedwards** [comment]
Are you able to please provide an example of this behaviour occuring?  I've tried to reproduce this using the PipeParser and each segment is only getting parsed once as far as I've determined ...

If you could provide a test case that would be great

**Comment by AFAde** [comment]
Hi, thanks for the quick answer. 
If you set a break point at the end of the private method nextToken(char[] localDelimiters) of the class Tokenizer and start debugging the parsing of any hl7 file, you will see the process hitting that break point twice for each line of the hl7 file.

If you cannot reproduce that, I will send you an hl7 file that causes me this issue

**Comment by duaneedwards** [comment]
Thanks AFAde, I've reproduced this behaviour and have implemented a simple caching system of the extracted token which will prevent the duplicate extraction of each subsequent token.

I'll push this change up in a new release to nuget later today.

Sorry it took a while, I just needed to convince myself this change actually does improve performance.  My findings were such that you wouldn't really see a benefit from this change until you start processing messages at volume.  The first couple message parses were taking very similar amounts of time, but after about 10 or so the runtime / JIT must have kicked in and then I could see a difference between two approaches.
